runtime/lib/capi/ExecutionContext.hpp

```diff
@@ -157,112 +177,232 @@ class SharedLibraryManager final {
+        device_pool;
+    size_t pool_counter{0}; // Counter for generating unique keys for devices
 
-        // Convert to a mutable string
-        std::string rtd_lib_str{rtd_lib};
+    bool initial_tape_recorder_status;
 
-        // LCOV_EXCL_START
-        // The following if-elif is required for C++ tests where these backend devices
-        // are linked in the interface library of the runtime. (check runtime/CMakeLists.txt)
-        // Besides, this provides support for runtime device (RTD) libraries added to the system
-        // path. This maintains backward compatibility for specifying a device using its name.
-        // TODO: This support may need to be removed after updating the C++ unit tests.
-        if (rtd_lib == "lightning.qubit" || rtd_lib == "lightning.kokkos") {
-            _device_name =
-                (rtd_lib == "lightning.qubit") ? "LightningSimulator" : "LightningKokkosSimulator";
-            update_device_pl2runtime(rtd_lib_str, "lightning");
-        }
-        else if (rtd_lib == "braket.aws.qubit" || rtd_lib == "braket.local.qubit") {
-            _device_name = "OpenQasmDevice";
-            update_device_pl2runtime(rtd_lib_str, "openqasm");
-        }
-        // LCOV_EXCL_STOP
+    // ExecutionContext pointers
+    std::unique_ptr<MemoryManager> memory_man_ptr{nullptr};
+    std::unique_ptr<PythonInterpreterGuard> py_guard{nullptr};
+
+  public:
+    explicit ExecutionContext() : initial_tape_recorder_status(false)
+    {
+        memory_man_ptr = std::make_unique<MemoryManager>();
+    };
+
+    ~ExecutionContext()
+    {
+        memory_man_ptr.reset(nullptr);
+        py_guard.reset(nullptr);
+    };
+
+    void setDeviceRecorderStatus(bool status) noexcept { initial_tape_recorder_status = status; }
+
+    [[nodiscard]] auto getDeviceRecorderStatus() const -> bool
+    {
+        return initial_tape_recorder_status;
+    }
+
+    [[nodiscard]] auto getMemoryManager() const -> const std::unique_ptr<MemoryManager> &
+    {
+        return memory_man_ptr;
+    }
+
+    [[nodiscard]] auto addDeviceToPool(std::shared_ptr<RTDeviceType> device) -> size_t
+    {
+        // Add a new device
+        const size_t key = pool_counter++;
+        device_pool[key] = std::make_pair(RTDeviceStatus::Init, device);
 
 #ifdef __build_with_pybind11
-        if (_device_name == "OpenQasmDevice" && !Py_IsInitialized()) {
-            _py_guard = std::make_unique<PythonInterpreterGuard>(); // LCOV_EXCL_LINE
+        if (!py_guard && device->getDeviceName() == "OpenQasmDevice" && !Py_IsInitialized()) {
+            py_guard = std::make_unique<PythonInterpreterGuard>(); // LCOV_EXCL_LINE
         }
 #endif
 
-        QuantumDevice *impl = loadDevice(rtd_lib_str);
-        _device_ptr.reset(impl);
-        return true;
+        return key;
     }
 
-    void releaseDevice() noexcept { _device_ptr.reset(nullptr); }
+    [[nodiscard]] auto getDevice(std::string_view rtd_lib, std::string_view rtd_name,
```

Comment:
It is more create or get, right?

Reply:
This method returns a device pointer with its unique index via either creating a new device or reuse a released device. As it doesn't always create a new device instance, I didn't call it `createDevice` and instead it's named `getDevice`. Let me know what you think.

Reply:
Minor: maybe ``getOrCreate`` would be nice

Reply:
Fixed!